ibm/resource_ibm_securitygroup_rule.go

```diff
@@ -0,0 +1,253 @@
+package ibm
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/softlayer/softlayer-go/datatypes"
+	"github.com/softlayer/softlayer-go/filter"
+	"github.com/softlayer/softlayer-go/services"
+	"github.com/softlayer/softlayer-go/sl"
+)
+
+func resourceIBMSecurityGroupRule() *schema.Resource {
+	return &schema.Resource{
+		Create:   resourceIBMSecurityGroupRuleCreate,
+		Read:     resourceIBMSecurityGroupRuleRead,
+		Delete:   resourceIBMSecurityGroupRuleDelete,
+		Exists:   resourceIBMSecurityGroupRuleExists,
+		Importer: &schema.ResourceImporter{},
+
+		Schema: map[string]*schema.Schema{
+			"direction": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				Description:  "Direction of rule: ingress or egress",
+				ValidateFunc: validateSecurityRuleDirection,
+			},
+			"ether_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "IP version IPv4 or IPv6",
+				Default:      "IPv4",
+				ForceNew:     true,
+				ValidateFunc: validateSecurityRuleEtherType,
+			},
+			"port_range_min": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+			"port_range_max": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+			"protocol": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "icmp, tcp or udp",
+				ForceNew:     true,
+				ValidateFunc: validateSecurityRuleProtocol,
+			},
+			"security_group_id": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func findMatchingRule(securitygroupid int, rule *datatypes.Network_SecurityGroup_Rule,
+	service services.Network_SecurityGroup) (*datatypes.Network_SecurityGroup_Rule, error) {
+
+	var filters []filter.Filter
+	if rule.PortRangeMax != nil {
+		filters = append(filters, filter.Path("rules.portRangeMax").Eq(rule.PortRangeMax))
+	}
+	if rule.PortRangeMin != nil {
+		filters = append(filters, filter.Path("rules.portRangeMin").Eq(rule.PortRangeMin))
+	}
+
+	filters = append(filters, filter.Path("rules.direction").Eq(rule.Direction))
+
+	if rule.Ethertype != nil {
+		filters = append(filters, filter.Path("rules.ethertype").Eq(rule.Ethertype))
+	}
+	if rule.Protocol != nil {
+		filters = append(filters, filter.Path("rules.protocol").Eq(rule.Protocol))
+	}
```

Comment:
These filters as they currently stand may call 2 rules identical that are not equivalent.

You may have a rule with port 22 open for TCP to 192.168.0.0/24, and another rule for port 22 TCP to remoteGroup 129347. These aren't the same but would be called so with this function.

You'll need to add filters for remoteIp and remoteGroup as well here.